rust-symcrypt/src/chacha.rs

```diff
@@ -0,0 +1,115 @@
+/* ChaChaPoly1305 Functions. For further documentation please refer to symcrypt.h */
+
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+pub fn chacha20_poly1305_encrypt(
+    key: &[u8],
+    nonce: &[u8],
+    auth_data: Option<&[u8]>,
+    src: &[u8],
+) -> Result<(Vec<u8>, [u8; 16]), SymCryptError> {
+    let (auth_data_ptr, auth_data_len) = auth_data.map_or_else(
+        || (std::ptr::null(), 0 as symcrypt_sys::SIZE_T),
+        |data| (data.as_ptr(), data.len() as symcrypt_sys::SIZE_T),
+    );
+
+    let mut dst = vec![0u8; src.len()];
+    let mut tag = [0u8; 16]; // ChaCha tag length must be 16 bytes
+
+    unsafe {
+        // SAFETY: FFI calls
+        match symcrypt_sys::SymCryptChaCha20Poly1305Encrypt(
+            key.as_ptr(),
+            key.len() as symcrypt_sys::SIZE_T,
+            nonce.as_ptr(),
+            nonce.len() as symcrypt_sys::SIZE_T,
+            auth_data_ptr,
+            auth_data_len,
+            src.as_ptr(),
+            dst.as_mut_ptr(),
+            dst.len() as symcrypt_sys::SIZE_T,
+            tag.as_mut_ptr(),
+            tag.len() as symcrypt_sys::SIZE_T,
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok((dst, tag)),
+            err => Err(err.into()),
+        }
+    }
+}
+
+pub fn chacha20_poly1305_decrypt(
+    key: &[u8],
+    nonce: &[u8],
+    auth_data: Option<&[u8]>,
+    src: &[u8],
+    tag: &[u8; 16], // ChaCha tag length must be 16 bytes
+) -> Result<Vec<u8>, SymCryptError> {
+    let (auth_data_ptr, auth_data_len) = auth_data.map_or_else(
+        || (std::ptr::null(), 0 as symcrypt_sys::SIZE_T),
+        |data| (data.as_ptr(), data.len() as symcrypt_sys::SIZE_T),
+    );
+
+    let mut dst = vec![0u8; src.len()];
+
+    unsafe {
+        // SAFETY: FFI calls
+        match symcrypt_sys::SymCryptChaCha20Poly1305Decrypt(
+            key.as_ptr(),
+            key.len() as symcrypt_sys::SIZE_T,
+            nonce.as_ptr(),
+            nonce.len() as symcrypt_sys::SIZE_T,
+            auth_data_ptr,
+            auth_data_len,
+            src.as_ptr(),
+            dst.as_mut_ptr(),
+            dst.len() as symcrypt_sys::SIZE_T,
+            tag.as_ptr(),
+            tag.len() as symcrypt_sys::SIZE_T,
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(dst),
+            err => Err(err.into()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Maybe add some tests that should return an error.

Reply:
will add